keps/sig-scheduling/2458-preferred-fit-strategy/README.md

```diff
@@ -0,0 +1,664 @@
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+Potentially reduce the number of allowed combinations too much. This can be mitigated
+by adding more configuration strategies whereever applicable. 
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+Define a new `PreferredFitStrategy` type as follows:
+
+```go
+
+type StrategyType string
+
+const (
+    LeastAllocated StrategyType = "LeastAllocated"
+    MostAllocated StrategyType = "MostAllocated"
+    BalancedAllocated StrategyType = "BalancedAllocated"
+    RequestedToCapacityRatio StrategyType = "RequestedToCapacityRatio"
+)
+
+type PreferredFitStrategy struct {
+	metav1.TypeMeta
+
+    // Strategy selects which strategy to run.
+    Strategy StrategyType
+    
+    // Resources to consider when scoring.
+    // The default resource set includes "cpu" and "memory" with an equal weight.
+	// Allowed weights go from 1 to 100.
+    Resources []ResourceSpec
+
+    // Arguemnts specific to RequestedToCapacityRatio strategy.
+    RequestedToCapacityRatio *RequestedToCapacityRatio
+}
+
+type RequestedToCapacityRatio struct {
+    // Points defining priority function shape
+    Shape []UtilizationShapePoint
+}
+
+// Note that the two types defined below already exist in the scheduler's component config API.
+// ResourceSpec represents a single resource.
+type ResourceSpec struct {
+	// Name of the resource.
+	Name string
+	// Weight of the resource.
+	Weight int64
+}
+
+// UtilizationShapePoint represents a single point of a priority function shape.
+type UtilizationShapePoint struct {
+	// Utilization (x axis). Valid values are 0 to 100. Fully utilized node maps to 100.
+	Utilization int32
+	// Score assigned to a given utilization (y axis). Valid values are 0 to 10.
+	Score int32
+}
+
+```
+
+Add `PreferredFitStrategy` to `NodeResourcesFitArgs`:
+
+```go
+
+// NodeResourcesFitArgs holds arguments used to configure the NodeResourcesFit plugin.
+type NodeResourcesFitArgs struct {
+	metav1.TypeMeta
+
+	// IgnoredResources is the list of resources that NodeResources fit filter
+	// should ignore.
+	IgnoredResources []string
+	// IgnoredResourceGroups defines the list of resource groups that NodeResources fit 
+    // filter should ignore.
+	// e.g. if group is ["example.com"], it will ignore all resource names that begin
+	// with "example.com", such as "example.com/aaa" and "example.com/bbb".
+	// A resource group name can't contain '/'.
+	IgnoredResourceGroups []string
+    
+    // PreferredFitStrategy selects the node resource scoring strategy.
+    PreferredFitStrategy *PreferredFitStrategy
+}
+```
+
+<!--
```

Comment:
Outline the deprecation strategy:

Legacy plugins can be used in v1beta1, but not v1beta2.